lib/ansible/module_utils/basic.py

```diff
@@ -2498,6 +2501,48 @@ def preserved_copy(self, src, dest):
         current_attribs = current_attribs.get('attr_flags', '')
         self.set_attributes_if_different(dest, current_attribs, True)
 
+    def lock_file(self, filename, lock_timeout=0):
+        if self.check_mode:
+            return True
+
+        self.lockfd = None
+        tmpdir = None
+        if not tmpdir:
+            tmpdir = tempfile.gettempdir()
+
+        lock_file = os.path.join(tmpdir, 'ansible-{0}.lock'.format(os.path.basename(filename)))
+        l_wait = 0.2
+        if lock_timeout > 0:
+            e_secs = 0
+            while e_secs < lock_timeout:
+                try:
+                    self.lockfd = open(lock_file, 'w')
+                    fcntl.flock(self.lockfd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+                    return True
+                except BlockingIOError:
+                    time.sleep(l_wait)
+                    e_secs += l_wait
+                    continue
+            return False
+
+        self.lockfd = open(lock_file, 'w')
+        fcntl.flock(self.lockfd, fcntl.LOCK_EX)
+        return True
+
+    def unlock_file(self):
+        if not self.lockfd or self.check_mode:
```

Comment:
I'm not sure if making this and lock_file conditional upon check_mode is the right thing to do.  A module might need to lock a file in order to read it and decide if changes should be made.

Reply:
add a `ignore_on_check_mode=False/true` option so the caller can decide, as it should know if it is needed or not

Reply:
check_mode is something that is AnsibleModule knowledge.  lock and unlock are about file operations.  So erally, a caller who only wants to lock when not in check_mode should do:
``` python
if not module.check_mode:
    lock_file(filepath)
```
check_mode should not be part of the lock_file() and unlock_file() API.

Reply:
I agree it's better to keep it out of (un)lock_file. Thanks for the clarification, wasn't sure what was preferred so I just searched for usage of check_mode in basic.py.